src/modules/0.0.19/aws_api_gateway/index.ts

```diff
@@ -0,0 +1,165 @@
+          // use the same ID as the one inserted, and set the name as is optionally returned
+          newApi.id = r.id;
+          newApi.name = r.name;
+          await this.module.api.db.update(newApi, ctx);
+          out.push(newApi);
+        }
+      }
+      return out;
+    },
+    read: async (ctx: Context, apiId?: string) => {
+      const client = (await ctx.getAwsClient()) as AWS;
+      if (apiId) {
+        const rawApi = await this.getApi(client.apiGatewayClient, apiId);
+        if (!rawApi) return undefined;
+        return this.apiMapper(rawApi);
+      } else {
+        const rawApis = (await this.getApis(client.apiGatewayClient)) ?? [];
+        const out = [];
+        for (const i of rawApis) {
+          const outApi = this.apiMapper(i);
+          if (outApi) out.push(outApi);
+        }
+        return out;
+      }
+    },
+    updateOrReplace: (a: Api, b: Api) => 'update',
+    update: async (rs: Api[], ctx: Context) => {
+      const client = (await ctx.getAwsClient()) as AWS;
+      const out = [];
+      for (const r of rs) {
+        const cloudRecord = ctx?.memo?.cloud?.Api?.[r.apiId ?? ''];
+        const isUpdate = Object.is(this.module.api.cloud.updateOrReplace(cloudRecord, r), 'update');
+        if (isUpdate) {
+          // api id is generated by aws, we cannot modify it
+          if (cloudRecord.protocolType !== r.protocolType) {
+            // restore
+            cloudRecord.id = r.id;
+            await this.module.api.db.update(cloudRecord, ctx);
+            out.push(cloudRecord);
+          } else {
+            const input: UpdateApiCommandInput = {
+              ApiId: r.apiId,
+              Name: r.name,
+              Description: r.description,
+              DisableExecuteApiEndpoint: r.disableExecuteApiEndpoint,
+              Version: r.version,
+            };
+            const res = await this.updateApi(client.apiGatewayClient, input);
+            if (res) {
+              const newApi = this.apiMapper(res);
+              newApi.name = r.name;
+              newApi.id = r.id;
+              // Save the record back into the database to get the new fields updated
+              await this.module.api.db.update(newApi, ctx);
+              out.push(newApi);
+            } else {
+              throw new Error('Error updating API');
+            }
+          }
+        }
+      }
+      return out;
+    },
+    delete: async (rs: Api[], ctx: Context) => {
+      const client = (await ctx.getAwsClient()) as AWS;
+      for (const r of rs) {
+        await this.deleteApi(client.apiGatewayClient, r.apiId);
+      }
+    },
+  });
+
+  constructor(module: AwsApiGatewayModule) {
+    super();
+    this.module = module;
+    super.init();
+  }
+
+  apiMapper(instance: any) {
+    const r: Api = new Api();
+    if (!instance.ApiId) throw new Error('Received an API without a id');
+    if (!instance.Name) throw new Error('Received an API without a name');
```

Comment:
here we are not `throw`ing but returning `undefined`, this way we do not fail and just ignore misconfigured resources.